# Bài tập 1 — Thu thập dữ liệu**Môn:** Lập trình Phân tích Dữ liệu (2101681)  **Nhóm 7, 8**---## Mục tiêuThu thập dữ liệu tin đăng nhà đất TP.HCM từ 3 nguồn:- CafeLand.vn- batdongsan.vn- chotot.com**Người phụ trách:** Nguyễn Bá Công, Hà Trọng Hữu Duy

In [7]:
import sys
from pathlib import Path
import pandas as pd
import numpy as np
sys.path.append(str(Path.cwd().parent))

pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 200)
print('done')

done


## 1. Kiểm tra file dữ liệu thôFile dữ liệu đặt trong `data/raw/`. Xem hướng dẫn tải tại `docs/DATA.md`.

In [2]:
from src.utils import doc_config, duong_dan_goc
cfg = doc_config()
thu_muc_tho = duong_dan_goc() / cfg['duong_dan']['du_lieu_tho']
print("Các file trong data/raw/:")
for f in sorted(thu_muc_tho.glob('*')):
    if f.is_file() and f.name != '.gitkeep':
        print(f"  {f.name:<35} {f.stat().st_size/1024/1024:>8.2f} MB")

Các file trong data/raw/:
  bat_dong_san_raw.xls                   51.30 MB
  cafeland_raw.csv                        9.58 MB


## 2. Đọc và xem thử từng nguồn

In [8]:
import importlib
import src.data.lam_sach_du_lieu as lsd

importlib.reload(lsd)

doc_cafeland = lsd.doc_cafeland
doc_batdongsan = lsd.doc_batdongsan
doc_chotot = lsd.doc_chotot

print("Đã reload code mới")

Đã reload code mới


In [9]:
from src.data.lam_sach_du_lieu import doc_cafeland, doc_batdongsan, doc_chotot
# Đọc nguồn CafeLand
df_cafeland = doc_cafeland(thu_muc_tho / cfg['file_nguon']['cafeland'])
df_cafeland.head()

  [CafeLand]    đọc được 9,883 dòng


,tieu_de,gia_ban,dien_tich_m2,tinh_thanh,quan_huyen,mo_ta_dac_diem,loai_hinh,nguon_du_lieu,url
0,Môi giới nhà đất Nguyễn Thành Phước 0966755***,NaN,74.8,TP. Hồ Chí Minh,NaN,NaN,NaN,cafeland,https://nhadat.cafeland.vn/moi-gioi/nguyen-tha...
1,Bán 2 lô đất biệt thự vườn có thể làm homestay...,3150.0,459.0,TP. Hồ Chí Minh,NaN,Bán 2 lô đất biệt thự vườn có thể làm homestay...,Biệt thự,cafeland,https://nhadat.cafeland.vn/ban-2-lo-dat-biet-t...
2,"Bán căn hộ 2PN 2WC 74,8m2 full NT The Sun Aven...",7000.0,74.8,TP. Hồ Chí Minh,NaN,"Căn hộ chung cư tại The Sun Avenue, Mai Chí Th...",Căn hộ,cafeland,https://nhadat.cafeland.vn/ban-can-ho-2pn-2wc-...
3,MẶT TIỀN NGUYỄN CÔNG TRỨ VS NAM KỲ- DT: 8X18- ...,140000.0,140.0,TP. Hồ Chí Minh,NaN,NHÀ 1 ĐỜI CHỦ CẦN BÁN CHIA TÀI SẢN CHO CON ĐỊN...,Bán nhà riêng,cafeland,https://nhadat.cafeland.vn/mat-tien-nguyen-con...
4,"Bán nhà mặt tiền ĐHT 05, Đông Hưng Thuận – 128...",13400.0,128.4,TP. Hồ Chí Minh,NaN,"Căn nhà có diện tích đất 128,4m², diện tích xâ...",Nhà mặt tiền,cafeland,https://nhadat.cafeland.vn/ban-nha-mat-tien-dh...


In [10]:
# Đọc nguồn batdongsan.vn

df_bds = doc_batdongsan(thu_muc_tho / 'bat_dong_san_raw.xls')
df_bds.head()
# cfg['file_nguon']['batdongsan']

  [batdongsan]  đọc được 30,149 dòng


,tieu_de,gia_ban,dien_tich_m2,tinh_thanh,quan_huyen,mo_ta_dac_diem,loai_hinh,nguon_du_lieu,url
0,"Bán nhà riêng tại 32 Huyền Quang, Phường Tân Đ...",11500.0,33.5,TP. Hồ Chí Minh,NaN,"Bán nhà riêng 32 Huyền Quang, Phường Tân Định,...",Nhà riêng,batdongsan,https://batdongsan.vn/ban-nha-rieng-tai-32-huy...
1,"CHỈ TỪ 1,98 TỶ SỞ HỮU CĂN HỘ THE EMERALD 68 - ...",1980.0,35.0,TP. Hồ Chí Minh,NaN,"💥 CHỈ TỪ 1,98 TỶ SỞ HỮU CĂN HỘ THE EMERALD 68 ...",Căn hộ,batdongsan,https://batdongsan.vn/chi-tu-198-ty-so-huu-can...
2,"Bán nhà riêng 32 Huyền Quang, Phường Tân Định,...",11500.0,33.5,TP. Hồ Chí Minh,NaN,"Bán nhà riêng 32 Huyền Quang, Phường Tân Định,...",Nhà riêng,batdongsan,https://batdongsan.vn/ban-nha-rieng-32-huyen-q...
3,"Bán đất mt đường nhựa Phan Thị So, xã Phước Vĩ...",7000.0,955.0,TP. Hồ Chí Minh,NaN,"Bán lô đất mặt tiền đường nhựa Phan Thị So , x...",Biệt thự,batdongsan,https://batdongsan.vn/ban-dat-mt-duong-nhua-ph...
4,750 Triệu - Có Thể Mở Khóa Căn Hộ 2PN2WC The A...,2500.0,67.0,TP. Hồ Chí Minh,NaN,750 Triệu - Có Thể Mở Khóa Căn Hộ 2PN2WC The A...,Căn hộ,batdongsan,https://batdongsan.vn/750-trieu-co-the-mo-khoa...


In [ ]:
# Đọc nguồn chotot.com# LƯU Ý: hàm này tự động trích lại diện tích từ tiêu đề# vì cột `area` gốc bị lỗi crawl (xem docs/DATA.md)df_chotot = doc_chotot(thu_muc_tho / cfg['file_nguon']['chotot'])df_chotot.head()

## 3. Tổng quan 3 nguồn

In [11]:
from src.utils import tom_tat_dataframe
for ten, df in [('CafeLand', df_cafeland), ('batdongsan.vn', df_bds)]:#, ('chotot.com', df_chotot)]:
      tom_tat_dataframe(df, ten)


TÓM TẮT: CafeLand
Kích thước      : 9,883 dòng × 9 cột
Trùng lặp toàn bộ: 0 dòng
Thiếu dữ liệu   :
    quan_huyen                       9,883 (100.0%)
    dien_tich_m2                     1,224 ( 12.4%)
    gia_ban                            900 (  9.1%)
    mo_ta_dac_diem                     422 (  4.3%)
    loai_hinh                          421 (  4.3%)

TÓM TẮT: batdongsan.vn
Kích thước      : 30,149 dòng × 9 cột
Trùng lặp toàn bộ: 0 dòng
Thiếu dữ liệu   :
    quan_huyen                      30,149 (100.0%)
    loai_hinh                        8,355 ( 27.7%)
    dien_tich_m2                        71 (  0.2%)
    gia_ban                             14 (  0.0%)


## 4. Nhận xét*(Viết nhận xét của nhóm tại đây)*- Nguồn nào có quy mô lớn nhất?- Nguồn nào có vấn đề chất lượng dữ liệu?- Bước tiếp theo cần xử lý gì?

Nguồn có quy mô lớn nhất: batdongsan.vn (30,149 dòng so với 9,883 dòng của CafeLand).

Nguồn có vấn đề chất lượng dữ liệu: Cả hai nguồn đều gặp lỗi nghiêm trọng về cấu trúc vị trí khi cột quan_huyen bị thiếu dữ liệu hoàn toàn (100%), đồng thời batdongsan.vn có tỷ lệ khuyết thiếu cột loai_hinh cao (27.7%), còn CafeLand thiếu nhiều thông tin dien_tich_m2 (12.4%) và gia_ban_trieu (9.1%).

Bước tiếp theo cần xử lý: Trích xuất/tách thông tin quận huyện và phường xã từ cột địa chỉ thô (nếu có), điền hoặc loại bỏ các dòng thiếu giá trị trọng yếu (gia_ban_trieu, dien_tich_m2), và chuẩn hóa các cột phân loại trước khi đưa vào mô hình dự đoán.